accepted/0017-add-funding-support.md

```diff
@@ -0,0 +1,138 @@
+# Add `funding` support to `package.json`
+
+## Summary
+
+This RFC identifies an initial means of tooling to describe & notify consumers of a package's monetary support information. `npm` has been [vocal about our commitment](https://blog.npmjs.org/post/187382017885/supporting-open-source-maintainers) to providing a means for package maintainers to more sustainably support their work & this reflects an initial step toward that goal. 
+
+## Motivation
+
+Package maintainers want to clearly indentify how their software is currently, or could be in the future, supported monetarily.
+
+## Detailed Explanation
+
+* See [Prior Art](#prior-art)
+* Provide a means to **reference** existing backers or funding opportunities in `package.json`
+* Provide a means to **define the type** of backing or funding opportunities in `package.json`
+* Provide a means to **notify** package consumers of backers or funding opportunites with the cli
+* Provide a means to **view** dependency's backers or funding opportunites with the cli
+* Provide a means to **view** a package's backers or funding opportunites
+
+##  Rationale & Alternatives
+* Rationale: This is a straightforward and easily implemented solution to take the first step toward supporting Open Source Package Maintainence sustainability
+* Rationale: Building this functionality into the package management layer is the only reliable way to push funding notifications to the vast majority of developers who already use npm
+* Alternative: Continue to rely on third-party, opt-in, tools & platforms that only support a fraction of the desired audience
+* Alternative: Continue debating a larger, more complex schema /w correspondingly more complex tooling
+
+## Implementation
+
+* Add a `funding` field to `package.json`
+  * supports a string URL or object with...
+  * keys that support arbitrary string identifiers
+    * ex. `"sponsor"`, `"sponsors"`, `"donations"`, `"bounties"`, `"contributors"`, `"patrons"` etc.
+  * values that represent a string URL or an array of URLs
+* Add notification at the end of output of package installation that references the number of packages with `funding` defined 
+  * ex. `23 packages are looking for funding. Run "npm fund" to find out more.`
```

Comment:
This should probably differentiate between top-level and transitive deps somehow - it’s far more important to fund your direct deps (who may in turn fund their direct deps) than to support some arbitrarily deep transitive dependency, imo.

Reply:
I disagree with this specific comment, nested dependencies already takes a big hit in terms of awareness and in many ways they can be as vital as the first-level ones to a given project

I would rather steer away from making an explicit differentiation between them for the initial implementation and see how the community reacts to that 😊

Reply:
@ruyadorno they should be visible, but transitive deps can be transparently refactored in and out all the time, so it's not the same thing as direct ones.

Reply:
I agree with @ruyadorno that we should certainly include transitive dependencies by default.  We could indicate the difference in the UI, but they need to be included.

Reply:
That is, I'm saying, while we could indicate which ones are direct deps and which ones are transitive, we should endeavor to _not_ do so in a way that steers people towards only funding their direct dependencies.

Reply:
@isaacs why? people only have control over their direct deps. As a maintainer, if I swap out my package A's internals from dep B to dep C, and the consumers of A have been funding B (perhaps without realizing that B is there because A needs it), then shouldn't C be getting that funding instead?

In other words, imo it should be on the developers of each dep to fund their own deps - otherwise maintainers may be incentivized to fork/inline rather than have deps themselves. This doesn't seem like a desired consequence.

Reply:
@ljharb Because our dep chains are deep, and in practice, it doesn't trickle down efficiently.

There isn't an ideal solution, but imagine the situation where a module A needs funding, and is only a direct dep of B (which absolutely critically depends on it!), and B (which doesn't seek funding) is used by the 50 most popular npm packages.

If we only show the funding nag for direct deps, then we show B's nag to one person.  If we show it for all deps, we expose it to 5M devs.  If I were the maintainer of B, writing code that is downloaded billions of times each month and depended on by every Fortune 1000 company, it'd seem pretty unfair that my request for funding only goes out to one person.

The ideal solution, of course, is for npm to act as a broker, using the insight we have about downloads and dependencies to fairly direct funds where they're needed.  But that is a longer-term idea, and fraught with its own complexities.

Reply:
Of course, you can replace "doesn't show the nag" with "shows the nag but in a way that nudges you away from donating", and draw a conclusion that only differs in degree.  Metadeps are deps.

Reply:
@isaacs that's all fair, and i very much wouldn't want to *hide* transitive dep funding info - i'm mainly just saying it should be differentiated, so that users aren't confused about where the info comes from.